### PR TITLE
Manual backport: Update go to 1.21.6 and use .go-version file everywhere to release/1.3.x

### PR DESCRIPTION
**Pull request by curtbushko**
Manual backport of #384 to release/1.3.x

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -70,12 +58,26 @@ jobs:
...
+              pkg_suffix: "-fips",
+            }
+          - {
+              goos: "linux",
+              goarch: "arm64",
+              gotags: "fips",
+              env: "CGO_ENABLED=1 GOEXPERIMENT=boringcrypto CC=aarch64-linux-gnu-gcc",
+              fips: "+fips1402",
+              pkg_suffix: "-fips",
+            }
```
Comment (zalimeni) [inline .github/workflows/build.yml:80]: We might not want to reformat in these backports bc it'll bork any future backports to these lines (though I'm down to reformat from `main` if you want)

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -237,35 +241,46 @@ jobs:
...
-
+          - {
+              target-name: "release-ubi",
+              registry-id: "631f805e0d15f623c5996c2e",
+            }
+          - {
+              target-name: "release-fips-ubi",
+              registry-id: "649cced3e093d1cc7f3d9b61",
+              fips: "+fips1402",
+            }
```
Comment (zalimeni) [inline .github/workflows/build.yml:265]: Same here re: backport formatting

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -373,12 +392,7 @@ jobs:
       - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
       - uses: actions/setup-go@fac708d6674e30b6ba41289acaab6d4b75aa0753 # v4.0.1
         with:
-          # pinning this to 1.20.5 because this issue in go-testcontainers occurs
-          # in 1.20.6 with the error "http: invalid Host header, host port waiting failed"
-          # https://github.com/testcontainers/testcontainers-go/issues/1359
-          # remove setting this when the above issue is fixed so that the job will
-          # just get the go version from .go-version.
-          go-version: 1.20.5
+          go-version: ${{ needs.get-go-version.outputs.go-version }}
```
Comment (zalimeni) [inline .github/workflows/build.yml:395]: I think we need a `needs` added above (did we miss it on the main PR?)

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -387,7 +401,7 @@ jobs:
         with:
           name: consul-${{matrix.server.version}}-integration-tests-output
           path: integration-tests/output/
-          if-no-files-found: 'error'
+          if-no-files-found: "error"
```
Comment (zalimeni) [inline .github/workflows/build.yml:404]: Same here re: backport formatting

**Review by zalimeni — APPROVED** [review]
Small change suggestions but otherwise LGTM; approving to unblock once those are closed.

File: .github/workflows/consul-dataplane-checks.yaml [hunk 5]
```diff
@@ -42,16 +32,12 @@ jobs:
       - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
       - uses: actions/setup-go@fac708d6674e30b6ba41289acaab6d4b75aa0753 # v4.0.1
         with:
-          # pinning this to 1.20.5 because this issue in go-testcontainers occurs
-          # in 1.20.6 with the error "http: invalid Host header, host port waiting failed"
-          # https://github.com/testcontainers/testcontainers-go/issues/1359
-          # remove setting this when the above issue is fixed so that the job will
-          # just get the go version from .go-version.
-          go-version: 1.20.5
+          go-version: ${{ needs.get-go-version.outputs.go-version }}
```
Comment (zalimeni) [inline .github/workflows/consul-dataplane-checks.yaml:35]: Missing `needs` here too I think

File: .github/workflows/build.yml (hunk 3 above)
Comment (curtbushko) [reply .github/workflows/build.yml:375]: Yes, it is missing in the main PR. I will fix.

**Pull request merged**